kore/src/Kore/Step/Simplification/Unify.hs

```diff
@@ -0,0 +1,1006 @@
+import Kore.Internal.Symbol
+import Kore.Internal.TermLike
+import Kore.Rewriting.RewritingVariable
+import Kore.Step.Simplification.ExpandAlias
+import Kore.Syntax.PatternF
+    ( Const (..)
+    )
+import Kore.TopBottom
+import Prelude.Kore
+
+data Unification
+    = UnifyInt !InternalInt !InternalInt
+    | UnifyBool !InternalBool !InternalBool
+    | UnifyString !InternalString !InternalString
+    | UnifyDomainValue !Sort !(TermLike RewritingVariableName) !Sort !(TermLike RewritingVariableName)
+    | UnifyStringLiteral !Text !Text
+    | EqualsAndEquals
+    | BytesDifferent !InternalBytes !InternalBytes
+    | UnifyFirstBottom !Sort
+    | VariableFunctionEquals !(ElementVariable RewritingVariableName)
+    | EqualInjectiveHeadsAndEquals !Symbol ![TermLike RewritingVariableName] !Symbol ![TermLike RewritingVariableName]
+    | MatchInjs !(Inj (TermLike RewritingVariableName)) !(Inj (TermLike RewritingVariableName))
+    | MatchInjApp !Symbol
+    | MatchAppInj !Symbol
+    | MatchAppHeads !Symbol !Symbol
+    | UnifyBoolAnd !(TermLike RewritingVariableName) !(TermLike RewritingVariableName)
+    | UnifyBoolOr !UnifyBoolOrArgs
+    | UnifyBoolNot !UnifyBoolNotArgs
+    | UnifyKequalsEq !UnifyKequalsEqArgs
+    | UnifyIntEq !UnifyIntEqArgs
+    | UnifyStringEq !UnifyStringEqArgs
+    | UnifyEqualsEndianness !UnifyEqualsEndiannessArgs
+    | UnifyEqualsSignedness !UnifyEqualsSignednessArgs
+    | UnifyNotInKeys !UnifyNotInKeysArgs
+    | DomainValueAndConstructorErrors1
+    | DomainValueAndConstructorErrors2
+    | UnifyOverload1 !UnifyOverload1Args
+    | UnifyOverload2 !UnifyOverload2Args
+    | UnifyOverload3 !UnifyOverload3Args
+    | UnifyOverload4 !UnifyOverload4Args
+    | UnifyOverload5 !UnifyOverload5Args
+    | UnifyOverload6 !UnifyOverload6Args
+    | UnifyEqualsMap1 !UnifyMapEqualsArgs
+    | UnifyEqualsMap2 !UnifyMapEqualsVarArgs
+    | UnifyEqualsMap3 !UnifyMapEqualsVarArgs
+    | UnifyMapBottom
+    | UnifyEqualsSet1 !UnifySetEqualsArgs
+    | UnifyEqualsSet2 !UnifySetEqualsVarArgs
+    | UnifyEqualsSet3 !UnifySetEqualsVarArgs
+    | UnifySetBottom
+    | UnifyEqualsList1
+    | UnifyEqualsList2
+    | UnifyEqualsList3 !UnifyEqualsList3Args
+    | UnifyEqualsList4 !UnifyEqualsList4Args
+    | UnifyEqualsList5 !UnifyEqualsList5Args
+    | UnifyBoolAnd1
+    | UnifyBoolAnd2
+    | UnifyBoolAnd3
+    | UnifyBoolAnd4
+    | UnifyVariableFunctionAnd1 !(ElementVariable RewritingVariableName)
+    | UnifyVariableFunctionAnd2 !(ElementVariable RewritingVariableName)
+    | UnifyIfThenElse !(IfThenElse (TermLike RewritingVariableName))
+    | UnifyExpandAlias !UnifyExpandAliasArgs
+    | UnifyIsFunctionPatterns
+
+matchInt
+    :: TermLike RewritingVariableName
+    -> TermLike RewritingVariableName
+    -> Maybe Unification
+matchInt first second
+    | InternalInt_ int1 <- first
+    , InternalInt_ int2 <- second
+        = Just $ UnifyInt int1 int2
+    | otherwise = Nothing
+{-# INLINE matchInt #-}
```

Comment:
Every one of these should be localized to the module where we implement unification for that type (in this case, `InternalInt`).